cmd/autoscaler/main.go

```diff
@@ -201,8 +164,81 @@ func main() {
 	case <-egCh:
 	case <-stopCh:
 	}
+}
+
+func setupLogger() (*zap.SugaredLogger, zap.AtomicLevel) {
+	loggingConfigMap, err := configmap.Load("/etc/config-logging")
+	if err != nil {
+		log.Fatalf("Error loading logging configuration: %v", err)
+	}
+	loggingConfig, err := logging.NewConfigFromMap(loggingConfigMap)
+	if err != nil {
+		log.Fatalf("Error parsing logging configuration: %v", err)
+	}
+	return logging.NewLoggerFromConfig(loggingConfig, component)
+}
+
+func setupChannels() (<-chan struct{}, chan *autoscaler.StatMessage) {
+	stopCh := signals.SetupSignalHandler()
+	statsCh := make(chan *autoscaler.StatMessage, statsBufferLen)
+	return stopCh, statsCh
+}
+
+func setupClients(logger *zap.SugaredLogger, stopCh <-chan struct{}) (kubernetes.Interface, clientset.Interface, scale.ScalesGetter) {
+	cfg, err := clientcmd.BuildConfigFromFlags(*masterURL, *kubeconfig)
+	if err != nil {
+		logger.Fatalw("Error building kubeconfig", zap.Error(err))
+	}
+
+	kubeClientSet, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalw("Error building kubernetes clientset", zap.Error(err))
+	}
+
+	if err := version.CheckMinimumVersion(kubeClientSet.Discovery()); err != nil {
+		logger.Fatalf("Version check failed: %v", err)
+	}
+
+	// This is based on how Kubernetes sets up its scale client based on discovery:
+	// https://github.com/kubernetes/kubernetes/blob/94c2c6c84/cmd/kube-controller-manager/app/autoscaling.go#L75-L81
+	restMapper := buildRESTMapper(kubeClientSet, stopCh)
+	scaleClient, err := scale.NewForConfig(cfg, restMapper, dynamic.LegacyAPIPathResolverFunc,
+		scale.NewDiscoveryScaleKindResolver(kubeClientSet.Discovery()))
+	if err != nil {
+		logger.Fatalw("Error building scale clientset", zap.Error(err))
+	}
+
+	servingClientSet, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalw("Error building serving clientset", zap.Error(err))
+	}
+
+	return kubeClientSet, servingClientSet, scaleClient
+}
+
+func setupAutoscalerConfig(logger *zap.SugaredLogger) *autoscaler.DynamicConfig {
+	rawConfig, err := configmap.Load("/etc/config-autoscaler")
+	if err != nil {
+		logger.Fatalw("Error reading autoscaler configuration", zap.Error(err))
+	}
+	dynConfig, err := autoscaler.NewDynamicConfigFromMap(rawConfig, logger)
+	if err != nil {
+		logger.Fatalw("Error parsing autoscaler configuration", zap.Error(err))
+	}
+	return dynConfig
+}
+
+func setupConfigWatcher(kc kubernetes.Interface, logger *zap.SugaredLogger, level zap.AtomicLevel, dynConfig *autoscaler.DynamicConfig) *configmap.InformedWatcher {
```

Comment:
This doesn't look like a coherent method. Takes random inputs for random config maps (level for logging, dynConfig for autoscaler, ...etc) and doesn't really reduce the code complexity. I find the old way to be more readable.

Reply:
Reverted.

This function is tricky, because I wanted to hide the setup details from the top level and have the helper both create a watcher and set it up to watch configs.  The inputs for creating a watcher is not related to the the setup inputs, as you pointed out.  Making the parameters coherent is possible, but it's lengthy.